### DISCUSSION TITLE
External macro implementation type could not be found

### DISCUSSION THREAD

**Discussion by MiguelG97** [pointfreeco/swift-composable-architecture#3419] [General]
### Description
Can someone guide me on how to set up my project to not have the following issues when implementing TCA?
[Github repo:](https://github.com/MiguelG97/voyansi_digital_twin)

External macro implementation type 'ComposableArchitectureMacros.ReducerMacro' could not be found for macro 'Reducer()

External macro implementation type 'ComposableArchitectureMacros.ObservableStateMacro' could not be found for macro 'ObservableState()'

External macro implementation type 'ComposableArchitectureMacros.ReducerMacro' could not be found for macro 'Reducer()'

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

xcode to not throw errors

### Actual behavior

xcode is throwing errors

### Reproducing project

[Github repo:](https://github.com/MiguelG97/voyansi_digital_twin)

All I did is installed the TCA package through SPM.
Once finished I added the library to my target project.
Then, when trying to implement the code a bunch of messages show up:

```
import ComposableArchitecture

@Reducer
struct LayoutSlice {
    
    @ObservableState
    struct State{
        var isLandScape = false
    }
    
    enum Action{
        case setLandScape
    }
    
    var body: some ReducerOf<Self>{
        Reduce{ state, action in
            switch action{
            case .setLandScape:
                state.isLandScape = true
                return .none
            }
        }
        
    }
}
```

### The Composable Architecture version information

1.15.0

### Destination operating system

IOS 17.5

**Comment by mbrandonw** [comment]
Hi @MiguelG97, this project builds just fine for me without any errors. There is a chance you have not enabled macros for your project. In the build errors for your project you may see a number of errors like this:

<img width="295" alt="Screenshot 2024-09-30 at 10 15 25 AM" src="https://github.com/user-attachments/assets/fa603482-d30b-4efa-9a92-18528b0f8225">

If you enable all of those it should work fine.

Since this isn't an issue with the library I am going to convert it to a discussion. Please feel free to continue the conversation over there!

**Reply by MiguelG97** [reply]
That's strange, I remind enabling all macros.
I appreciate your response anyways

**Comment by JeromeTonnelierOgury** [comment]
Hi!
For some reasons, I've got the same behavior, but the weirdest thing is that the project compiles and works. 
I tried to remove TCA and add it again (so that I can enable macros when Xcode asks me to), but the errors keep coming back. Basically, after doing that, if I don't touch anything, I have no error. But as soon as I add event just a line break in a reducer, all errors comes back :-/ 
I tried to remove all cache/derived data and start again, same result.
What's rellay annoying is that when there is an actual error, I just can't find it. One example os error would be `...MobileMenuView/MobileMenuFeature.swift:24:8 External macro implementation type 'ComposableArchitectureMacros.ReducerMacro' could not be found for macro 'Reducer()'; No such file or directory`
Any idea ? It's a pretty big project with Cocoapods, SMP (both local and github hosted).
Thanks 🙏

**Reply by JeromeTonnelierOgury** [reply]
And it does this only with this project of course, no issues with TCA per se :-)

**Reply by mbrandonw** [reply]
Hi @JeromeTonnelierOgury, do you have custom build configurations? For example, something other than the default Debug and Release configurations? That can cause this problem. 

Personally I don't think custom configurations are useful because it seems that Xcode has a lot of problems with it, but if it's too much trouble to change back then you could at least try adding back the default Debug and Release ones. Though I'm not sure if that's enough to get the project back into good shape.

**Reply by JeromeTonnelierOgury** [reply]
Hi and thanks for your answer. No, no custom xcconfig, and still have Debug/Release. I also added a PreProd configuration fyi, do you think it can be an issue ?

**Reply by holzerryan** [reply]
This was my issue when I encountered this error.  SPM does not support custom configurations.  Debug and Release are the only configurations that work for SPM (for now).  To solve this problem, I created environment variables in the scheme and then only used custom configurations for CI build pipelines.  In my situation, the custom configuration was required for separate bundle identifiers.  @mbrandonw pointed me in this direction and solved my problems.

Also note...the errors that state the macro Reducer() cannot be found are only in the Xcode UI.  The project did compile for me, but it made debugging any real issues next to impossible.

**Reply by holzerryan** [reply]
Here's a reference to my closed issue.
https://github.com/pointfreeco/swift-composable-architecture/discussions/3401#discussion-7212031

**Reply by JeromeTonnelierOgury** [reply]
Thanks a lot, I'll have a look at it :-)

**Reply by usiddiqui-google** [reply]
For us, the issue turned out to using Swift 6 in the swift-syntax package while our project was still on Swift 5. So updating it from `package(url: "https://github.com/swiftlang/swift-syntax.git", exact: "600.0.1")` to `.package(url: "https://github.com/swiftlang/swift-syntax.git", exact: "510.0.2")`, followed by deleting derived data and rebuilding packages fixed the issue.

**Reply by usiddiqui-google** [reply]
@mbrandonw lmk if its worth adding it here? I will do a PR.
https://pointfreeco.github.io/swift-composable-architecture/main/documentation/composablearchitecture/reducer/#Error-External-macro-implementation--could-not-be-found

**Reply by mbrandonw** [reply]
Hi @usiddiqui-google, that seems very surprising. I believe this is an Xcode bug and that the version of Swift Syntax can't really have any impact on this. I would be happy to update the docs, but first I would want to see conclusive evidence that it does indeed fix the problem.

**Reply by davidmorra** [reply]
> Hi @JeromeTonnelierOgury, do you have custom build configurations? For example, something other than the default Debug and Release configurations? That can cause this problem.
> 
> Personally I don't think custom configurations are useful because it seems that Xcode has a lot of problems with it, but if it's too much trouble to change back then you could at least try adding back the default Debug and Release ones. Though I'm not sure if that's enough to get the project back into good shape.

I had the exact same problem with the latest version. Everything works and builds, but that macro errors don't disappear. Indeed problem was in the custom configuration. Once i switched to the default configuration errors disappear. 
Thanks!

**Reply by medef00** [reply]
> For us, the issue turned out to using Swift 6 in the swift-syntax package while our project was still on Swift 5. So updating it from `package(url: "https://github.com/swiftlang/swift-syntax.git", exact: "600.0.1")` to `.package(url: "https://github.com/swiftlang/swift-syntax.git", exact: "510.0.2")`, followed by deleting derived data and rebuilding packages fixed the issue.

Doesn't work for me. Project Swift 6, swift-syntax 600.0.1 🥲

**Reply by vadamask** [reply]
> Hi @JeromeTonnelierOgury, do you have custom build configurations? For example, something other than the default Debug and Release configurations? That can cause this problem.
> 
> Personally I don't think custom configurations are useful because it seems that Xcode has a lot of problems with it, but if it's too much trouble to change back then you could at least try adding back the default Debug and Release ones. Though I'm not sure if that's enough to get the project back into good shape.

Work for me, thanks

**Comment by danscan** [comment]
For posterity: After fighting with this issue for HOURS, I was creating a simple repro project to post here, and I figured out **it goes away if you just do a CMD+B build** 🙃

**Reply by mbrandonw** [reply]
The main reason this happens is because the macros have not been built in order for Xcode to understand what `@Reducer` means. So it's always a good idea to do a full build before working on the app.

**Reply by VladimirCreator** [reply]
It really helped…

**Comment by rt12379** [comment]
I am trying to build a project with TCA using Azure pipelines, but the build fails with the same error.
The strange thing is that on Xcode I get the same error but can still build.
The project is on swift 5 but the method of setting the swift-syntax version to 510.0.2 did not work.
I am using Custom configuration, which I think is fine for the CI build pipeline. Has anyone solved this error on Azure pipelines?

**Comment by attilamarosi** [comment]
I had the same errors in a brand-new Xcode project (26.2, Swift 5). I saw some warnings in the Issue Navigator for Macros, and after tapping each warning and selecting “Trust and Enable” in the popup, all the errors were gone. The app now builds and runs correctly.

**Comment by sitnikovdev** [comment]
When I switch to the tab with build information, the macro errors disappear. When I return to the file editor, the errors appear again. However, after pressing  CMD+B , the project builds successfully and the app launches without any issues. But the errors still remain after the build.

<img width="480" height="201" alt="fantom" src="https://github.com/user-attachments/assets/9e3b7b7a-aca8-4efc-96d2-59d56915d467" />

All lines with the same error:
 "External macro implementation type 'ComposableArchitectureMacros.ReducerMacro' could not be found for macro 'Reducer()'; plugin for module 'ComposableArchitectureMacros' "
 
Using stack:
•	XCode 26.6
•	SwiftUI / Swfit 6.3
•	TCA 1.25
•	Tuist
